### ISSUE TITLE
Mysterious NameError inside Sidekiq

### ISSUE THREAD

**Issue by DmitryTsepelev** [fxn/zeitwerk#150]
Hi, thank you for a great gem!

I'm facing a strange issue somewhere between Rails, Sidekiq and Zeitwerk, please move the issue if it won't sound related. Imagine a following structure:

```
app/
  services/
    analytics_events/
      google_analytics.rb
      whatever_else_analytics.rb
   analytics_events.rb
workers/
  user/
    sign_in_worker.rb
```

`AnalyticsEvents` service has a list of methods representing events we want to send to one or more analytics services, which are stored inside `analytics_events/`:

```ruby
module AnalyticsEvents
  module_function

  def registration(*)
    Ga.registration(*)
    # ...
  end
end

class AnalyticsEvents::Ga
  module_function

  def registration(*)
     # ...
  end
end
```

Various Sidekiq workers (e.g., `User::SignInWorker`, but the problem appears in different places) call methods of `AnalyticsEvents` and sometimes `NameError: uninitialized constant AnalyticsEvents::Ga` (or any other constant inside `AnalyticsEvents`) is raised. We have autoscaling, so by "sometimes" I mean "sometimes new Sidekiq starts and errors happen inside this exact process, but all other processes work just fine".

I've turned on logging, but it seems to work fine:

```
Zeitwerk@rails.main: autoload set for AnalyticsEvents, to be autovivified from /workdir/app/services/analytics_events
Zeitwerk@rails.main: autoload set for AnalyticsEvents, to be loaded from /workdir/app/services/analytics_events.rb 
```

I've tried to catch that `NameError` and take a look inside `AnalyticsEvents` module: no surprizes, it's empty. Any ideas what I should try next?

**Comment by fxn** [comment]
How strange.

Is the application eager loading in those environments where it fails?

**Comment by DmitryTsepelev** [comment]
Yes, `config.eager_load = true` is set in production environment.

**Comment by fxn** [comment]
It is indeed surprising. Let me ask some questions:

1) Would it be possible to get Zeitwerk logs when that happens?
2) If the exception is raised, `AnalyticsEvents` does not have children constants, does it have the expected methods?
3) In development mode, do you see [this warning](https://github.com/rails/rails/blob/6143b1121f14da6cf5690f1821fb4b0db1eaa8e8/railties/lib/rails/application/finisher.rb#L51-L75) for any constant?

**Comment by DmitryTsepelev** [comment]
> 1. Would it be possible to get Zeitwerk logs when that happens?

Sure, just tell me what exactly we need to check

> 2. If the exception is raised, AnalyticsEvents does not have children constants, does it have the expected methods?

Forgot to mention! I've tried to log it and there are no methods in the `AnalyticsEvents`, looks like it loaded an empty `AnalyticsEvents` and nothing else. The only place where `AnalyticsEvents` is defined is `app/services/analytics_events.rb` and it's used as a namespace inside `app/services/analytics_events/*.rb`.

> 3. In development mode, do you see this warning for any constant? 

No errors

Also, forgot to mention versions I run: Rails 6.1, Sidekiq 6.0.7, Zeitwerk 2.4.2

**Comment by fxn** [comment]
> Sure, just tell me what exactly we need to check

Ideally everything from Zeitwerk (better grep everything else out for privacy). Since this is strange, the more info we have the better, maybe we can spot something off in the logs. Please feel free to send them privately to me at fxn@hashref.com.

**Comment by fxn** [comment]
We're looking into this out of band. We've seen so far that eager loading seems to eager load `app/services` just fine, except for some reason those specific analytics files are not reported as eager loaded, but they got `autoload` calls. I suspect there is some kind of external interference of some kind, but it has to be seen.

We're adding traces to have more data to help debugging, but we need to wait for the error to happen, which is not predictable.

**Comment by fxn** [comment]
We've seen the root cause is a gem loading application code while Rails is still booting. We'll deploy a workaround and we'll give it a couple of days to confirm.

**Comment by DmitryTsepelev** [comment]
I've deployed a fix on Thursday and it seems to fix the problem, so I'm closing the issue.

The problem was caused by a gem that loaded some of my workers _in the separate thread_ (!) as soon as it was required (in Gemfile). One of the workers referenced a class inside `AnalyticsEvents`, which resulted to a situation, when there was only one class that was loaded from a module and Zeitwerk could not load the rest because it seemed like the module was already loaded.

The solution was to delay gem initialisation until Rails is fully loaded (i.e., require gem in the `config.after_initialize` block). Also, watch your threads! 🙂

A huge thanks to @fxn who volunteered to read my enormously huge logs and debug the code without even seeing it, I have no idea how it would take me to realize that there is a different thread messing up my initialization process.
